File: proyecto-titulo-2025/apps/vecindApp/frontend/src/app/services/espacio.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse, HttpHeaders } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { 
  EspacioCreateRequest, 
  EspacioResponse, 
  EspacioListResponse,
  ApiError 
} from '../interfaces/espacio.interface';

@Injectable({
  providedIn: 'root'
})
export class EspacioService {
  private readonly API_URL = 'http://localhost:8000/api'; // URL base del backend

  constructor(private http: HttpClient) {}

  /**
   * Crea un nuevo espacio comunitario
   */
  createEspacio(espacioData: EspacioCreateRequest): Observable<EspacioResponse> {
    const headers = this.getAuthHeaders();
    
    return this.http.post<EspacioResponse>(`${this.API_URL}/espacios/`, espacioData, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  /**
   * Crea un nuevo espacio comunitario con archivo
   */
  createEspacioWithFile(espacioData: EspacioCreateRequest, file: File): Observable<EspacioResponse> {
    const formData = new FormData();
    
    // Agregar todos los campos del espacio al FormData
    formData.append('nombre', espacioData.nombre);
    formData.append('tipo', espacioData.tipo);
    formData.append('capacidad', espacioData.capacidad.toString());
    formData.append('valor', espacioData.valor.toString());
    formData.append('max_horas', espacioData.max_horas.toString());
    formData.append('activo', espacioData.activo.toString());
    formData.append('id_junta', espacioData.id_junta.toString());
    
    // Agregar arrays si existen
    if (espacioData.permitido && espacioData.permitido.length > 0) {
      espacioData.permitido.forEach(item => formData.append('permitido', item));
    }
    
    if (espacioData.no_permitido && espacioData.no_permitido.length > 0) {
      espacioData.no_permitido.forEach(item => formData.append('no_permitido', item));
    }
    
    // Agregar el archivo
    formData.append('foto', file, file.name);
    
    const headers = this.getAuthHeadersForFile();
    
    return this.http.post<EspacioResponse>(`${this.API_URL}/espacios/`, formData, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  /**
   * Obtiene un espacio por su ID
   */
  getEspacioById(id: number): Observable<EspacioResponse> {
    return this.http.get<EspacioResponse>(`${this.API_URL}/espacios/${id}`)
      .pipe(
        catchError(this.handleError)
      );
  }

  /**
   * Obtiene todos los espacios de una junta
   */
  getEspaciosByJunta(
    idJunta: number, 
    activoOnly: boolean = true,
    pagina: number = 1,
    porPagina: number = 10
  ): Observable<EspacioListResponse> {
    const params = new URLSearchParams({
      activo_only: activoOnly.toString(),
      pagina: pagina.toString(),
      por_pagina: porPagina.toString()
    });

    return this.http.get<EspacioListResponse>(`${this.API_URL}/espacios/junta/${idJunta}?${params}`)
      .pipe(
        catchError(this.handleError)
      );
  }

  /**
   * Actualiza un espacio existente
   */
  updateEspacio(id: number, espacioData: Partial<EspacioCreateRequest>): Observable<EspacioResponse> {
    const headers = this.getAuthHeaders();
    
    return this.http.put<EspacioResponse>(`${this.API_URL}/espacios/${id}`, espacioData, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  /**
   * Elimina un espacio (soft delete)
   */
  deleteEspacio(id: number): Observable<void> {
    const headers = this.getAuthHeaders();
    
    return this.http.delete<void>(`${this.API_URL}/espacios/${id}`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  /**
   * Obtiene los headers de autenticación
   */
  private getAuthHeaders(): HttpHeaders {
    const token = localStorage.getItem('token');
    return new HttpHeaders({
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`
    });
  }

  /**
   * Obtiene los headers de autenticación para archivos (sin Content-Type)
   */
  private getAuthHeadersForFile(): HttpHeaders {
    const token = localStorage.getItem('token');
    return new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
  }

  /**
   * Maneja errores HTTP
   */
  private handleError = (error: HttpErrorResponse): Observable<never> => {
    let errorMessage = 'Error desconocido';
    
    if (error.error instanceof ErrorEvent) {
      // Error del lado del cliente
      errorMessage = `Error: ${error.error.message}`;
    } else {
      // Error del lado del servidor
      if (error.error && error.error.detail) {
        errorMessage = error.error.detail;
      } else {
        errorMessage = `Error ${error.status}: ${error.statusText}`;
      }
    }
    
    console.error('Error en EspacioService:', errorMessage);
    return throwError(() => new Error(errorMessage));
  };
}

File: proyecto-titulo-2025/apps/vecindApp/backend/src/services/espacio_service.py
"""
Servicio para manejo de espacios comunitarios.

Maneja la creación, actualización y consulta de espacios comunitarios.
"""

import logging
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import selectinload
from typing import Optional, List, Dict, Any
from decimal import Decimal

from src.database.models.espacio import Espacio
from src.database.models.junta import Junta
from src.schemas.espacio_schemas import (
    EspacioCreateRequest,
    EspacioUpdateRequest,
    EspacioResponse,
    EspacioListResponse
)

logger = logging.getLogger(__name__)


class EspacioService:
    """
    Servicio para manejar espacios comunitarios.
    """

    def __init__(self, db: AsyncSession):
        """Inicializa el servicio con la sesión de base de datos."""
        self.db = db

    async def create_espacio(
        self, 
        espacio_data: EspacioCreateRequest,
        user_id: int
    ) -> EspacioResponse:
        """
        Crea un nuevo espacio comunitario.
        
        Args:
            espacio_data: Datos del espacio a crear
            user_id: ID del usuario que crea el espacio
            
        Returns:
            EspacioResponse: Datos del espacio creado
            
        Raises:
            ValueError: Si la junta no existe o el usuario no tiene permisos
            IntegrityError: Si hay conflicto de datos
        """
        try:
            # 1. Verificar que la junta existe
            junta = await self._get_junta_by_id(espacio_data.id_junta)
            if not junta:
                raise ValueError(f"Junta con ID {espacio_data.id_junta} no encontrada")

            # 2. Verificar que el usuario tiene permisos para crear espacios en esta junta
            # TODO: Implementar verificación de permisos basada en roles
            # Por ahora, asumimos que cualquier usuario autenticado puede crear espacios

            # 3. Crear el espacio
            nuevo_espacio = Espacio(
                id_junta=espacio_data.id_junta,
                nombre=espacio_data.nombre,
                tipo=espacio_data.tipo,
                capacidad=espacio_data.capacidad,
                valor=espacio_data.valor,
                foto=espacio_data.foto,
                permitido=espacio_data.permitido,
                no_permitido=espacio_data.no_permitido,
                max_horas=espacio_data.max_horas,
                activo=espacio_data.activo
            )

            self.db.add(nuevo_espacio)
            await self.db.commit()
            await self.db.refresh(nuevo_espacio)

            logger.info(f"Espacio '{nuevo_espacio.nombre}' creado exitosamente con ID {nuevo_espacio.id_espacio}")

            return EspacioResponse.from_orm(nuevo_espacio)

        except IntegrityError as e:
            await self.db.rollback()
            logger.error(f"Error de integridad al crear espacio: {e}")
            raise ValueError("Error al crear el espacio. Verifique que los datos sean válidos.")
        except Exception as e:
            await self.db.rollback()
            logger.error(f"Error inesperado al crear espacio: {e}")
            raise

    async def get_espacio_by_id(self, espacio_id: int) -> Optional[EspacioResponse]:
        """
        Obtiene un espacio por su ID.
        
        Args:
            espacio_id: ID del espacio
            
        Returns:
            EspacioResponse o None si no existe
        """
        result = await self.db.execute(
            select(Espacio).where(Espacio.id_espacio == espacio_id)
        )
        espacio = result.scalar_one_or_none()
        
        if espacio:
            return EspacioResponse.from_orm(espacio)
        return None

    async def get_espacios_by_junta(
        self, 
        id_junta: int, 
        activo_only: bool = True,
        pagina: int = 1,
        por_pagina: int = 10
    ) -> EspacioListResponse:
        """
        Obtiene todos los espacios de una junta.
        
        Args:
            id_junta: ID de la junta
            activo_only: Si solo mostrar espacios activos
            pagina: Número de página (1-indexed)
            por_pagina: Elementos por página
            
        Returns:
            EspacioListResponse: Lista paginada de espacios
        """
        # Construir query base
        query = select(Espacio).where(Espacio.id_junta == id_junta)
        
        if activo_only:
            query = query.where(Espacio.activo == True)
        
        # Contar total
        count_query = select(Espacio).where(Espacio.id_junta == id_junta)
        if activo_only:
            count_query = count_query.where(Espacio.activo == True)
        
        total_result = await self.db.execute(count_query)
        total = len(total_result.scalars().all())
        
        # Aplicar paginación
        offset = (pagina - 1) * por_pagina
        query = query.offset(offset).limit(por_pagina)
        
        # Ejecutar query
        result = await self.db.execute(query)
        espacios = result.scalars().all()
        
        # Convertir a response
        espacios_response = [EspacioResponse.from_orm(espacio) for espacio in espacios]
        
        return EspacioListResponse(
            espacios=espacios_response,
            total=total,
            pagina=pagina,
            por_pagina=por_pagina
        )

    async def update_espacio(
        self, 
        espacio_id: int, 
        espacio_data: EspacioUpdateRequest,
        user_id: int
    ) -> Optional[EspacioResponse]:
        """
        Actualiza un espacio existente.
        
        Args:
            espacio_id: ID del espacio a actualizar
            espacio_data: Datos a actualizar
            user_id: ID del usuario que actualiza
            
        Returns:
            EspacioResponse: Datos del espacio actualizado o None si no existe
            
        Raises:
            ValueError: Si el usuario no tiene permisos
        """
        try:
            # 1. Verificar que el espacio existe
            espacio = await self._get_espacio_by_id(espacio_id)
            if not espacio:
                return None

            # 2. Verificar permisos (TODO: implementar verificación de roles)
            
            # 3. Preparar datos para actualización
            update_data = {}
            for field, value in espacio_data.dict(exclude_unset=True).items():
                if value is not None:
                    update_data[field] = value

            if not update_data:
                # No hay nada que actualizar
                return EspacioResponse.from_orm(espacio)

            # 4. Actualizar
            await self.db.execute(
                update(Espacio)
                .where(Espacio.id_espacio == espacio_id)
                .values(**update_data)
            )
            await self.db.commit()

            # 5. Obtener el espacio actualizado
            updated_espacio = await self._get_espacio_by_id(espacio_id)
            logger.info(f"Espacio {espacio_id} actualizado exitosamente")

            return EspacioResponse.from_orm(updated_espacio)

        except Exception as e:
            await self.db.rollback()
            logger.error(f"Error al actualizar espacio {espacio_id}: {e}")
            raise

    async def delete_espacio(self, espacio_id: int, user_id: int) -> bool:
        """
        Elimina un espacio (soft delete - lo marca como inactivo).
        
        Args:
            espacio_id: ID del espacio a eliminar
            user_id: ID del usuario que elimina
            
        Returns:
            bool: True si se eliminó exitosamente
            
        Raises:
            ValueError: Si el usuario no tiene permisos
        """
        try:
            # 1. Verificar que el espacio existe
            espacio = await self._get_espacio_by_id(espacio_id)
            if not espacio:
                return False

            # 2. Verificar permisos (TODO: implementar verificación de roles)
            
            # 3. Soft delete - marcar como inactivo
            await self.db.execute(
                update(Espacio)
                .where(Espacio.id_espacio == espacio_id)
                .values(activo=False)
            )
            await self.db.commit()

            logger.info(f"Espacio {espacio_id} eliminado (marcado como inactivo)")
            return True

        except Exception as e:
            await self.db.rollback()
            logger.error(f"Error al eliminar espacio {espacio_id}: {e}")
            raise

    async def _get_espacio_by_id(self, espacio_id: int) -> Optional[Espacio]:
        """Obtiene un espacio por ID (método privado)."""
        result = await self.db.execute(
            select(Espacio).where(Espacio.id_espacio == espacio_id)
        )
        return result.scalar_one_or_none()

    async def _get_junta_by_id(self, junta_id: int) -> Optional[Junta]:
        """Obtiene una junta por ID (método privado)."""
        result = await self.db.execute(
            select(Junta).where(Junta.id_junta == junta_id)
        )
        return result.scalar_one_or_none()
